src/modules/lua_mtev.c

```diff
@@ -5286,10 +5293,208 @@ mtev_lua_deserialize(lua_State *L, const lua_data_t *data){
+nl_shared_queues_dequeue(const char *key, size_t key_len) {
+  void *vsq;
+  struct shared_queue *sq = NULL;
+  struct shared_queue_node *node = NULL;
+  pthread_mutex_lock(&shared_queues_mutex);
+  if(!mtev_hash_retrieve(&shared_queues, key, key_len, &vsq)) {
+    sq = calloc(1, sizeof(*sq));
+    mtev_hash_store(&shared_queues, strdup(key), key_len, sq);
+  }
+  else {
+    sq = vsq;
+  }
+
+  if(sq->head) {
+    node = sq->head;
+    sq->head = sq->head->next;
+    if(sq->head == NULL) sq->tail = NULL;
+  }
+  pthread_mutex_unlock(&shared_queues_mutex);
+
+  return node;
+}
+
+static int
+nl_shared_waitfor_wakeup(eventer_t e, int mask, void *closure, struct timeval *now) {
+  (void)mask;
+  struct shared_queues_wait *crutch = (struct shared_queues_wait *)closure;
+  struct shared_queue_node *node = nl_shared_queues_dequeue(crutch->key, strlen(crutch->key));
+
+  if(!crutch->ci) {
+    mtev_lua_eventer_cl_cleanup(e);
+  }
+  else {
+    mtev_lua_deregister_event(crutch->ci, e, 0);
+    if(node) {
+      if(node->data) {
+        mtev_lua_deserialize(crutch->L, node->data);
+        mtev_lua_free_data((lua_data_t *)node->data);
+      }
+      else {
+        lua_pushnil(crutch->L);
+      }
+      mtev_lua_lmc_resume(crutch->ci->lmc, crutch->ci, 1);
+    }
+    else {
+      if(crutch->forever) {
+        crutch->deadline = *now;
+        crutch->deadline.tv_sec += 10000;
+      }
+      if(compare_timeval(*now, crutch->deadline) < 0) {
+        crutch->e = eventer_alloc_timer(nl_shared_waitfor_wakeup, crutch, &crutch->deadline);
+        mtev_lua_register_event(crutch->ci, crutch->e);
+        eventer_add(crutch->e);
+      }
+      else {
+        lua_pushnil(crutch->L);
+        mtev_lua_lmc_resume(crutch->ci->lmc, crutch->ci, 1);
+      }
+    }
+    free(node);
   }
+  shared_queues_wait_free(crutch);
+  return 0;
 }
-/*! \lua mtev.shared_set()
+
+static int
+nl_shared_waitfor(lua_State *L) {
+  struct shared_queue_node *node = NULL;
+  size_t key_len;
+  const char *key;
+  double timeout = -1;
+  if(lua_gettop(L) > 2 || !lua_isstring(L,1) || (lua_gettop(L) == 2 && !lua_isnumber(L,2)))
+    return luaL_error(L, "bad parameters to mtev.shared_waitfor(str [, timeout])");
+  key = lua_tolstring(L, 1, &key_len);
+  if(lua_gettop(L) == 2) {
+    timeout = lua_tonumber(L, 2);
+  }
+
+  node = nl_shared_queues_dequeue(key, key_len);
+
+  if(node) {
+    if(node->data) {
+      mtev_lua_deserialize(L, (lua_data_t *)node->data);
+      mtev_lua_free_data((lua_data_t *)node->data);
+    }
+    else {
+      lua_pushnil(L);
+    }
+    free(node);
+    return 1;
+  }
+
+  /* There is no item yes, so we need to wait for one assuming timeout is non-zero */
+  if(timeout == 0) return 0;
+
+  mtev_lua_resume_info_t *ci;
+  ci = mtev_lua_find_resume_info(L, mtev_true); /* we cannot suspect without context */
+
+  struct shared_queues_wait *crutch = calloc(1, sizeof(*crutch));
+  double ntimeout = timeout < 0 ? 10900 : timeout;
```

Comment:
Where does the value `10900` here come from?

Reply:
Ha. Actually a typo, supposed to be 10000. But it is just an arbitrarily large timeout.

Reply:
resolved in comment.